prow/interrupts/interrupts.go

```diff
@@ -0,0 +1,206 @@
+	}
+	go handleInterrupt()
+}
+
+type manager struct {
+	// only one signal handler should be installed, so we use a cond to
+	// broadcast to workers that an interrupt has occurred
+	c *sync.Cond
+	// we record whether we've broadcast in the past
+	seenSignal bool
+	// we want to ensure that all registered servers and workers get a
+	// change to gracefully shut down
+	wg sync.WaitGroup
+}
+
+// handleInterrupt turns an interrupt into a broadcast for our condition.
+// This must be called _first_ before any work is registered with the
+// manager, or there will be a deadlock.
+func handleInterrupt() {
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt, syscall.SIGTERM, syscall.SIGABRT)
+	s := <-sig
+	logrus.WithField("signal", s).Info("Received signal.")
+	single.c.L.Lock()
+	single.seenSignal = true
+	single.c.Broadcast()
+	single.c.L.Unlock()
+}
+
+// wait executes the cancel when an interrupt is seen or if one has already
+// been handled
+func wait(cancel func()) {
+	single.c.L.Lock()
+	if !single.seenSignal {
+		single.c.Wait()
+	}
+	single.c.L.Unlock()
+	cancel()
+}
+
+// WaitForGracefulShutdown waits until all registered servers and workers
+// have had time to gracefully shut down, or times out. This function is
+// blocking.
+func WaitForGracefulShutdown() {
+	finished := make(chan struct{})
+	go func() {
+		single.wg.Wait()
+		close(finished)
+	}()
+	select {
+	case <-finished:
+		logrus.Info("All workers gracefully terminated, exiting.")
+	case <-time.After(1 * time.Minute):
+		logrus.Warn("Timed out waiting for workers to gracefully terminate, exiting.")
+	}
+}
+
+// Context returns a context that stays is cancelled when an interrupt hits.
+// Using this context is a weak guarantee that your work will finish before
+// process exit as callers cannot signal that they are finished. Prefer to use
+// Run().
+func Context() context.Context {
+	ctx, cancel := context.WithCancel(context.Background())
+	single.wg.Add(1)
+	go wait(func() {
+		cancel()
+		single.wg.Done()
+	})
+
+	return ctx
+}
+
+// Run will do work until an interrupt is received, then signal the
+// worker. This function is not blocking. Callers are expected to exit
+// only after WaitForGracefulShutdown returns to ensure all workers have
+// had time to shut down. This is preferable to getting the raw Context
+// as we can ensure that the work is finished before releasing our share
+// of the wait group on shutdown.
+func Run(work func(ctx context.Context)) {
+	ctx, cancel := context.WithCancel(context.Background())
+	single.wg.Add(1)
+	go func() {
+		defer single.wg.Done()
+		work(ctx)
+	}()
+
+	go wait(cancel)
+}
+
+// ListenAndServe runs the HTTP server and handles shutting it down
+// gracefully on interrupts. This function is not blocking. Callers
+// are expected to exit only after WaitForGracefulShutdown returns to
+// ensure all servers have had time to shut down.
+func ListenAndServe(server *http.Server, gracePeriod time.Duration) {
```

Comment:
Isn't `server.ListenAndServe()` blocking? If so I feel like this is a bad name.

Reply:
Why?

Reply:
Like, reading how the callers of the `interrupts` library use it, nothing seems wrong to me. I guess it could be surprising that it's not blocking, but that is also kind of the entire point of the library, so ...

Reply:
So give the function a different name.

Reply:
You are changing a blocking function to a non-blocking function. That is an important change to the function's behavior. Important functional changes should not be 100% transparent.

Which part of this reasoning do you disagree with?